src/ControlSystem/FunctionVsTime.hpp

```diff
@@ -0,0 +1,169 @@
+
+/// Function implementations
+template <size_t MaxDeriv>
+FunctionVsTimeNthDeriv<MaxDeriv>::FunctionVsTimeNthDeriv(
+    double t, value_type initial_func_and_derivs) noexcept
+    : deriv_info_{{t, std::move(initial_func_and_derivs)}} {}
+
+template <size_t MaxDeriv>
+typename FunctionVsTimeNthDeriv<MaxDeriv>::value_type
+FunctionVsTimeNthDeriv<MaxDeriv>::operator()(const double t) const noexcept {
+
+  const auto k = closest_time_iter(t);
+  const double dt = t - k->time;
+  const value_type derivs = k->derivs;
+  // initialize the return variable
+  // (with derivs, since gets overwritten anyways)
+  value_type result = derivs;
+
+  // the outer size is equal to the number of components
+  // the inner size is MaxDeriv+1
+  for (size_t i = 0; i < derivs.size(); i++) {
+    for (size_t d = 0; d < MaxDeriv; ++d) {
+      result[i][d] = derivs[i][MaxDeriv];
+      for (size_t dd = MaxDeriv; dd-- > d;) {
+        // TODO: benchmark test precomputing 1/ (dd-d+1), index by dd-d
+        result[i][d] *= dt / (dd - d + 1);
+        result[i][d] += derivs[i][dd];
+      }
+    }
+    result[i][MaxDeriv] = derivs[i][MaxDeriv];
+  }
+
+  return result;
+}
+
+template <size_t MaxDeriv>
+void FunctionVsTimeNthDeriv<MaxDeriv>::update(
+    double time_of_update,
+    const std::vector<double>& updated_max_deriv) noexcept {
+
+  if (time_of_update <= deriv_info_.back().time) {
+    ERROR("FunctionVsTimeNthDeriv::update: "
+          "t must be increasing from call to call.");
+  }
+
+  // get the current values, before updating the `MaxDeriv'th deriv
+  value_type func = this->operator()(time_of_update);
+
+  if (updated_max_deriv.size() != func.size()) {
+    ERROR("FunctionVsTimeNthDeriv::update: "
+          "the number of components trying to be updated ("
+          << updated_max_deriv.size() << ") does "
+          "not match the number of components ("
+          << func.size() << ") in the fvt.");
+  }
+
+  // update the MaxDerivs
+  for (size_t i = 0; i < updated_max_deriv.size(); i++) {
+    func[i][MaxDeriv] = updated_max_deriv[i];
+  }
+  deriv_info_.emplace_back(std::move(time_of_update),std::move(func));
+}
+
+// Return iterator pointing to k, such that times[k] <= t < times[k+1].
+// For t < times[0], return times.begin()
+// For t > times[times.size()-1] : return prev(times.end(),1)
+template <size_t MaxDeriv>
+auto FunctionVsTimeNthDeriv<MaxDeriv>::closest_time_iter(
+    const double t) const noexcept {
+  // this function now assumes that times is sorted, which is enforced
+  // by the update function.
+
+  // returns iterator pointing to the first element that is > than value.
+  // returns times.end() if no element > than value.
+  // if value < min(times), then returns times.begin() index.
+  auto it = std::upper_bound(deriv_info_.begin(), deriv_info_.end(), t,
+                       [](double t0, const DerivInfo& d) {return d.time > t0;});
+
+  // this section needs to change if extrapolation is not allowed.
```

Comment:
We think it is better to remove the extrapolation stuff (and always let the function be evaluated after the final time) . It will be the responsibility of the other code to determine whether or not it is appropriate to evaluate the function after the last time.  We will want a member function that will return the pair of times corresponding to the first and last times in deriv_info_